### PR TITLE
add completedTransaction as a subscription

### PR DESCRIPTION
**Pull request by eauge**
Implement a completedTransaction that can handle transactions from confidential contracts

### CODE REVIEW

**Comment by armaniferrante** [comment]
To summarize offline conversations, we need an answer to the following question: can we guarantee that the web3c.js client has access to a transaction hash prior to actually sending it off to the gateway.

Imagine the following situation

- Client sends transaction through provider (e.g., metamask). 
- The provider, e.g. metamask or hdwalletprovider signs the transaction, and then sends it simultaneously to the gateway and back to the web3c.js client.
- At this point the web3c.js subscribes to the gateway waiting for the transaction to be completed. But now we have a race condition. There's might not be a guarantee the subscription happens *before* the transaction is complete.

If this is indeed a problem, then the implementation needs to support the case where the subscription happens after the transaction is sent.

If we're using a websocket connection, one way around this is to subscribe to all transactions from a given address. Send the transaction through the exact same websocket connection, and then receive the response. This works because we're connected to the same gateway instance. If we can't guarantee that we're connected to the same gateway instance, the return value is not retrieved by the gateway that didn't initiate the transaction.

A more complicated solution would be to somehow store and allow the gateway to fetch the return value. This seems like the most immediate solution.

@willscott @peterjgilbert  thoughts here?

**Comment by eauge** [comment]
@armaniferrante good summary.

I would suggest to try to find a way to access the transactionHash of a transaction before sending it. If that's not possible, I would do the trick of subscribing to transactions for a `fromAddress`. 

I would avoid including storage to this unless strictly necessary. Then there would be all sorts of other problems.

File: web3c/oasis.js [hunk 1]
```diff
@@ -169,6 +171,75 @@ class Oasis {
...
+      }
+    });
+
+    completedTransaction.setRequestManager(options.web3._requestManager);
+    completedTransaction.attachToObject(this._oasisExclusiveSubscriptions);
+
+    this._subscribeCompletedTransaction = function(filter) {
+      const address = filter && filter.address ? filter.address : null;
+      if (address) {
+        delete filter.address;
```
[the remark this answers is not in the document]
Comment (eauge) [reply web3c/oasis.js]: sure let me add a comment. Reason is that `address` is not a filter property expected by the gateway, and the gateway will not accept payloads that have unexpected properties

File: web3c/oasis.js [hunk 2]
```diff
@@ -169,6 +171,75 @@ class Oasis {
   }
 
   _setupSubscribe(options) {
+    this._oasisExclusiveSubscriptions = {};
+    const completedTransaction = new Subscriptions({
+      name: 'subscribe',
+      type: 'eth',
+      subscriptions: {
+        completedTransaction: {
+          subscriptionName: 'completedTransaction', // replace subscription with this name
```
[the remark this answers is not in the document]
Comment (eauge) [reply web3c/oasis.js]: no, removed

[the remark this answers is not in the document]
Comment (eauge) [reply web3c/oasis.js]: forgot to push :)

File: web3c/oasis.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (eauge) [reply web3c/oasis.js]: forgot to push my comment :)

**Pull request merged**